moveit_ros/moveit_servo/src/utils/common.cpp

```diff
@@ -341,16 +341,16 @@ double jointLimitVelocityScalingFactor(const Eigen::VectorXd& velocities,
 }
 
 std::vector<int> jointsToHalt(const Eigen::VectorXd& positions, const Eigen::VectorXd& velocities,
-                              const moveit::core::JointBoundsVector& joint_bounds, double margin)
+                              const moveit::core::JointBoundsVector& joint_bounds, const std::vector<double>& margins)
 {
   std::vector<int> joint_idxs_to_halt;
   for (size_t i = 0; i < joint_bounds.size(); i++)
   {
     const auto joint_bound = (joint_bounds[i])->front();
     if (joint_bound.position_bounded_)
     {
-      const bool negative_bound = velocities[i] < 0 && positions[i] < (joint_bound.min_position_ + margin);
-      const bool positive_bound = velocities[i] > 0 && positions[i] > (joint_bound.max_position_ - margin);
+      const bool negative_bound = velocities[i] < 0 && positions[i] < (joint_bound.min_position_ + margins[i]);
+      const bool positive_bound = velocities[i] > 0 && positions[i] > (joint_bound.max_position_ - margins[i]);
```

Comment:
It's critical that `margins` have the correct dimension, otherwise this will crash and be difficult to debug. I'd suggest adding a dimension check within the `validateParams()` function of `servo.cpp`

Reply:
This ought to do it. In `servo.cpp`:

```
  if (servo_params.joint_limit_margins.size() !=
      robot_state->getJointModelGroup(servo_params.move_group_name).getActiveVariableCount())
  {
    RCLCPP_ERROR(logger_, "The joint_limits_margin array has the wrong size.");
    params_valid = false;
  }
```

Reply:
Thanks for pointing this out :)